tests/test_convert_text_to_mds.py

```diff
@@ -0,0 +1,198 @@
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+import sys
+
+import pytest
+
+# Add repo root to path so we can import scripts and test it
+repo_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
+sys.path.append(repo_dir)
+import pathlib
+from glob import glob
+from multiprocessing.pool import Pool
+from typing import Callable, Iterable, List, Optional
+from unittest.mock import Mock, patch
+
+import numpy as np
+from streaming import StreamingDataset
+from torch.utils.data import DataLoader, Dataset
+from transformers import AutoTokenizer
+
+from scripts.data_prep.convert_text_to_mds import (download_and_convert,
+                                                   is_already_processed, main,
+                                                   merge_shard_groups,
+                                                   write_done_file)
+
+
+class MockObjectStore():
+
+    def __init__(self, remote_folder: str, n_text_files: int,
+                 text_content: str):
+        os.makedirs(remote_folder, exist_ok=True)
+        for i in range(n_text_files):
+            with open(os.path.join(remote_folder, f'test{i}.txt'), 'w') as f:
+                f.write((text_content + ' ') * 500)
+        self.remote_folder = remote_folder
+        self.n_text_files = n_text_files
+
+    def download_object(self,
+                        object_name: str,
+                        filename: str,
+                        overwrite: bool = False):
+        with open(
+                os.path.join(self.remote_folder, os.path.basename(object_name)),
+                'rb') as remote_file, open(filename, 'wb') as local_file:
+            local_file.write(remote_file.read())
+
+    def list_objects(self, prefix: str) -> List[str]:
+        return glob(os.path.join(self.remote_folder, '*.txt'))
+
+    def upload_object(self, object_name: str, filename: str):
+        with open(
+                os.path.join(self.remote_folder, os.path.basename(object_name)),
+                'wb') as remote_file, open(filename, 'rb') as local_file:
+            remote_file.write(local_file.read())
+
+
+def _call_convert_text_to_mds(processes: int, tokenizer_name: str) -> None:
+    main(
+        tokenizer_name=tokenizer_name,
+        output_folder=f's3://fake-test-output-path',
+        input_folder=f's3://fake-test-input-path',
+        concat_tokens=2048,
+        eos_text='',
+        bos_text='',
+        no_wrap=False,
+        max_workers=1,
+        compression='zstd',
+        processes=processes,
+        args_str='Namespace()',
+        reprocess=False,
+    )
+
+
+# Mock starmap with no multiprocessing
+def _mock_starmap(func: Callable, args: Iterable):
```

Comment:
any reason this actually needs to be mocked?

Reply:
multi processing unfortunately messes with the function call counting with mocking

Reply:
hmm, i think i might know how to get around that after taking a second look... let's see

Reply:
yeah, that's going to be tricky to get working with nice mocking/call counts because memory isn't shared between processes. So we could count the calls by writing things to files instead of using the nice Mock.call_count if it's really important for us to test with full multiprocessing. wdyt?

Reply:
ah thats ok. we can rely on manual test for this.